optuna/terminator/improvement/emmr.py

```diff
@@ -0,0 +1,387 @@
+            X=normalized_params,
+            Y=standarized_score_vals,
+            is_categorical=(search_space.scale_types == gp_search_space.ScaleType.CATEGORICAL),
+            log_prior=prior.default_log_prior,
+            minimum_noise=prior.DEFAULT_MINIMUM_NOISE_VAR,
+            initial_kernel_params=kernel_params_t1,
+            deterministic_objective=self._deterministic,
+        )
+
+        theta_t_star_index = int(np.argmax(standarized_score_vals))
+        theta_t1_star_index = int(np.argmax(standarized_score_vals[:-1]))
+        theta_t_star = normalized_params[theta_t_star_index, :]
+        theta_t1_star = normalized_params[theta_t1_star_index, :]
+
+        cov_t_between_theta_t_star_and_theta_t1_star = _compute_gp_posterior_cov_two_thetas(
+            search_space,
+            normalized_params,
+            standarized_score_vals,
+            kernel_params_t,
+            theta_t_star_index,
+            theta_t1_star_index,
+        )
+
+        mu_t1_theta_t_with_nu_t, sigma_t1_theta_t_with_nu_t = _compute_gp_posterior(
+            search_space,
+            normalized_params[:-1, :],
+            standarized_score_vals[:-1],
+            normalized_params[-1, :],
+            kernel_params_t,
+            # Use kernel_params_t instead of kernel_params_t1.
+            # Use "t" under the assumption that "t" and "t1" are approximately the same.
+            # This is because kernel should same when computing KLD.
+            # For detailed information, please see section 4.4 of the paper:
+            # https://proceedings.mlr.press/v206/ishibashi23a/ishibashi23a.pdf
+        )
+        _, sigma_t_theta_t1_star = _compute_gp_posterior(
+            search_space,
+            normalized_params,
+            standarized_score_vals,
+            theta_t1_star,
+            kernel_params_t,
+        )
+        mu_t_theta_t_star, sigma_t_theta_t_star = _compute_gp_posterior(
+            search_space,
+            normalized_params,
+            standarized_score_vals,
+            theta_t_star,
+            kernel_params_t,
+        )
+        mu_t1_theta_t1_star, _ = _compute_gp_posterior(
+            search_space,
+            normalized_params[:-1, :],
+            standarized_score_vals[:-1],
+            theta_t1_star,
+            kernel_params_t1,
+        )
+
+        y_t = standarized_score_vals[-1]
+        kappa_t1 = _compute_standardized_regret_bound(
+            kernel_params_t1,
+            search_space,
+            normalized_params[:-1, :],
+            standarized_score_vals[:-1],
+            self._delta,
+            rng=self._rng.rng,
+        )
+
+        theorem1_delta_mu_t_star = mu_t1_theta_t1_star - mu_t_theta_t_star
+
+        alg1_delta_r_tilde_t_term1 = theorem1_delta_mu_t_star
+
+        theorem1_v = math.sqrt(
+            max(
+                1e-100,  # Note: sys.float_info.min can cause overflow in scipy_stats.
+                sigma_t_theta_t_star**2
```

Comment:
I think this sigma is the variance, so isn't `**2` wrong?

Reply:
I think that the following diff resolves this problem.

```bash
~/optuna$ git diff
diff --git a/optuna/terminator/improvement/emmr.py b/optuna/terminator/improvement/emmr.py
index 0361ed8bf..6ac3c572e 100644
--- a/optuna/terminator/improvement/emmr.py
+++ b/optuna/terminator/improvement/emmr.py
@@ -241,9 +241,9 @@ class EMMREvaluator(BaseImprovementEvaluator):
         theorem1_v = math.sqrt(
             max(
                 1e-100,  # Note: sys.float_info.min can cause overflow in scipy_stats.
-                sigma_t_theta_t_star**2
+                sigma_t_theta_t_star
                 - 2.0 * cov_t_between_theta_t_star_and_theta_t1_star
-                + sigma_t_theta_t1_star**2,
+                + sigma_t_theta_t1_star,
             )
         )
         theorem1_g = (mu_t_theta_t_star - mu_t1_theta_t1_star) / theorem1_v
@@ -252,15 +252,15 @@ class EMMREvaluator(BaseImprovementEvaluator):
         alg1_delta_r_tilde_t_term3 = theorem1_v * theorem1_g * scipy_stats.norm.cdf(theorem1_g)
 
         _lambda = prior.DEFAULT_MINIMUM_NOISE_VAR**-1
-        eq4_rhs_term1 = 0.5 * math.log(1.0 + _lambda * sigma_t1_theta_t_with_nu_t**2)
+        eq4_rhs_term1 = 0.5 * math.log(1.0 + _lambda * sigma_t1_theta_t_with_nu_t)
         eq4_rhs_term2 = (
-            -0.5 * sigma_t1_theta_t_with_nu_t**2 / (sigma_t1_theta_t_with_nu_t**2 + _lambda**-1)
+            -0.5 * sigma_t1_theta_t_with_nu_t / (sigma_t1_theta_t_with_nu_t + _lambda**-1)
         )
         eq4_rhs_term3 = (
             0.5
-            * sigma_t1_theta_t_with_nu_t**2
+            * sigma_t1_theta_t_with_nu_t
             * (y_t - mu_t1_theta_t_with_nu_t) ** 2
-            / (sigma_t1_theta_t_with_nu_t**2 + _lambda**-1) ** 2
+            / (sigma_t1_theta_t_with_nu_t + _lambda**-1) ** 2
         )
 
         alg1_delta_r_tilde_t_term4 = kappa_t1 * math.sqrt(
```

Reply:
I tried to reproduce the figure 4 in the paper.
With the above fix, the results are closer to those in the paper.

```python
import math
import matplotlib.pyplot as plt
import optuna
from optuna.terminator import EMMREvaluator


def holder_table(trial):
    x = trial.suggest_float("x", -10, 10)
    y = trial.suggest_float("y", -10, 10)
    return -abs(math.sin(x) * math.cos(y) * math.exp(abs(1 - math.sqrt(x**2 + y**2) / math.pi)))

def cross_in_tray(trial):
    x = trial.suggest_float("x", -10, 10)
    y = trial.suggest_float("y", -10, 10)
    return -0.0001 * (abs(math.sin(x) * math.sin(y) * math.exp(abs(100 - math.sqrt(x**2 + y**2) / math.pi))) + 1) ** 0.1

def easom(trial):
    x = trial.suggest_float("x", -10, 10)
    y = trial.suggest_float("y", -10, 10)
    return -math.cos(x) * math.cos(y) * math.exp(-((x - math.pi)**2 + (y - math.pi)**2))

def six_hump_camel(trial):
    x = trial.suggest_float("x", -10, 10)
    y = trial.suggest_float("y", -10, 10)
    return (4 - 2.1 * x**2 + x**4 / 3) * x ** 2 + x * y + (-4 + 4 * y**2) * y**2

def rosenbrock(trial):
    a = 1
    b = 100
    x = trial.suggest_float("x", -10, 10)
    y = trial.suggest_float("y", -10, 10)
    return (a - x) ** 2 + b * (y - x ** 2) ** 2

def booth(trial):
    x = trial.suggest_float("x", -10, 10)
    y = trial.suggest_float("y", -10, 10)
    return (x + 2 * y - 7) ** 2 + (2 * x + y - 5) ** 2


for objective, index, name in (
        (holder_table, "a", "holder table"),
        (cross_in_tray, "b", "cross in tray"),
        (easom, "c", "easom"),
        (six_hump_camel, "d", "six hump camel"),
        (rosenbrock, "e", "rosenbrock"),
        (booth, "f", "booth")):
    x = []
    y1 = []
    y2 = []
    y3 = []
    y4 = []
    y5 = []

    sampler = optuna.samplers.TPESampler(seed=0)
    study = optuna.create_study(sampler=sampler, study_name=name)
    emmr_improvement_evaluator = EMMREvaluator()

    for i in range(500):
        trial = study.ask()
        value = objective(trial)
        study.tell(trial, value)

        print(i, study.best_value, value, trial.params)

        ret = emmr_improvement_evaluator.evaluate(
            trials=study.trials,
            study_direction=study.direction,
        )

        if isinstance(ret, tuple):
            t1, t2, t3, t4, t5 = ret
            x.append(i + 1)
            y1.append(t1)
            y2.append(t2)
            y3.append(t3)
            y4.append(t4)
            y5.append(t5)

    plt.plot(x, y1, ".", label="abs(term_1)")
    plt.plot(x, y2, ".", label="term_2+term_3")
    plt.plot(x, y3, ".", label="term_4")
    plt.plot(x, y4, ".", label="kappa")
    plt.plot(x, y5, ".", label="term_4/kappa")
    plt.yscale("log")
    plt.ylim(1e-12, 1000)
    plt.legend()
    plt.title(f"({index}) {name}")
    plt.savefig(f"emmr_{index}_{name.replace(' ', '_')}.png")
    plt.clf()
```

I modified `emmr_improvement_evaluator.evaluate` to return the terms of the evaluation.

```python
        return (
            abs(alg1_delta_r_tilde_t_term1),
            alg1_delta_r_tilde_t_term2+alg1_delta_r_tilde_t_term3,
            alg1_delta_r_tilde_t_term4,
            kappa_t1,
            alg1_delta_r_tilde_t_term4/kappa_t1,
        )
```

* current PR
![emmr_e_rosenbrock](https://github.com/user-attachments/assets/8a2a96ee-c103-4ca0-9020-fe07cb5bdbe0)

* fixed
![emmr_e_rosenbrock](https://github.com/user-attachments/assets/393d43cd-9626-4546-885d-e2a1fa52195d)